tests/test_emulator_object_metadata.py

```diff
@@ -0,0 +1,126 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit test for Object Metadata operations in emulator.py."""
+
+import json
+import os
+import unittest
+from werkzeug.test import create_environ
+
+import emulator
+import gcs
+import testbench
+
+
+class TestEmulatorObjectMetadata(unittest.TestCase):
+    def setUp(self):
+        emulator.db = testbench.database.Database.init()
+        self.client = emulator.server.test_client()
+        # Avoid magic buckets in the test
+        os.environ.pop("GOOGLE_CLOUD_CPP_STORAGE_TEST_BUCKET_NAME", None)
+
+    def test_object_crud(self):
+        response = self.client.post(
+            "/storage/v1/b", data=json.dumps({"name": "bucket-name"})
+        )
+        self.assertEqual(response.status_code, 200)
+
+        payload = "The quick brown fox jumps over the lazy dog"
+        response = self.client.put(
+            "/bucket-name/fox.txt",
+            content_type="text/plain",
+            data=payload,
+        )
+        self.assertEqual(response.status_code, 200)
+
+        response = self.client.get("/storage/v1/b/bucket-name/o/fox.txt")
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue(
+            response.headers.get("content-type").startswith("application/json")
+        )
+        get_rest = json.loads(response.data)
+        expected_fields = {
+            "bucket": "bucket-name",
+            "name": "fox.txt",
+            "size": "%d" % len(payload),
+        }
+        self.assertEqual(get_rest, get_rest | expected_fields)
+
+        response = self.client.get(
+            "/storage/v1/b/bucket-name/o/fox.txt", query_string={"alt": "media"}
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data.decode("utf-8"), payload)
+
+        response = self.client.get(
+            "/storage/v1/b/bucket-name/o/fox.txt", query_string={"alt": "invalid"}
+        )
+        self.assertEqual(response.status_code, 500)
+
+        # Test PATCH by adding some metadata
+        metadata = {"key0": "label0"}
+        response = self.client.patch(
+            "/storage/v1/b/bucket-name/o/fox.txt",
+            data=json.dumps({"metadata": metadata}),
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue(
+            response.headers.get("content-type").startswith("application/json")
+        )
+        patch_rest = json.loads(response.data)
+        self.assertEqual(
+            patch_rest.get("metadata"), patch_rest.get("metadata") | metadata
```

Comment:
Could you help explain what this assertion does? 🙏

Reply:
Apparently `.assertDictContainsSubset(subset, full)` is deprecated (I get a warning).  Instead of do `full == full | subset` (or `subset | full`).  If `subset` is really a subset, then `|` adds nothing and the `==` succeeds :shrug:

Reply:
Clever and thanks for the explanation!